### PR TITLE
[JENKINS-37438] Allow to mark items as built-in

### PR DESCRIPTION
**Pull request by Vlatombe**
https://issues.jenkins-ci.org/browse/JENKINS-37438
@reviewbybees

### CODE REVIEW

File: core/src/main/java/jenkins/install/InstallUtil.java [hunk 1]
```diff
@@ -204,6 +208,21 @@ private static InstallState getDefaultInstallState() {
...
+        // the version is not persisted and returns 1.0, so try to check if
+        // they actually did anything
+        Map<String, TopLevelItem> items = new HashMap<String, TopLevelItem>(j.getItemMap());
+        for (Iterator<Map.Entry<String, TopLevelItem>> it = items.entrySet().iterator(); it.hasNext();) {
+            Map.Entry<String, TopLevelItem> entry = it.next();
+            if (new File(entry.getValue().getRootDir(), BUILT_IN_MARKER_FILE).exists()) {
+                it.remove();
+            }
+        }
+        return !items.isEmpty() || !j.getNodes().isEmpty();
```
Comment (cyrille-leclerc) [inline core/src/main/java/jenkins/install/InstallUtil.java]: Why does this "marker" not apply to Build Nodes as it applies to top level items?

File: core/src/main/java/jenkins/install/InstallUtil.java [hunk 2]
```diff
@@ -204,6 +208,21 @@ private static InstallState getDefaultInstallState() {
...
 
+    private static boolean isEdgeCaseUpgrade() {
+        Jenkins j = Jenkins.getInstance();
+        // Edge case: used Jenkins 1 but did not save the system config page,
+        // the version is not persisted and returns 1.0, so try to check if
+        // they actually did anything
+        Map<String, TopLevelItem> items = new HashMap<String, TopLevelItem>(j.getItemMap());
+        for (Iterator<Map.Entry<String, TopLevelItem>> it = items.entrySet().iterator(); it.hasNext();) {
+            Map.Entry<String, TopLevelItem> entry = it.next();
+            if (new File(entry.getValue().getRootDir(), BUILT_IN_MARKER_FILE).exists()) {
```
Comment (cyrille-leclerc) [inline core/src/main/java/jenkins/install/InstallUtil.java]: How do I create this `.builtIn` file on the file system of the Jenkins Home if I have create my "TopLevelItem" with a groovy script? For example in an init.groovy script?

File: core/src/main/java/jenkins/install/InstallUtil.java (hunk 1 above)
Comment (Vlatombe) [inline core/src/main/java/jenkins/install/InstallUtil.java]: Nodes are not `Item`s, they don't have a root directory on master.

File: core/src/main/java/jenkins/install/InstallUtil.java (hunk 2 above)
Comment (Vlatombe) [inline core/src/main/java/jenkins/install/InstallUtil.java]:
````
Item item = ...
new File(item.getRootDir(), ".builtIn").createNewFile()
````

File: core/src/main/java/jenkins/install/InstallUtil.java (hunk 1 above)
Comment (Vlatombe) [inline core/src/main/java/jenkins/install/InstallUtil.java]: However it is a good point, we should have a way to address build nodes as well.

Comment (Vlatombe) [inline core/src/main/java/jenkins/install/InstallUtil.java]: Actually nodes have a root dir but this is currently an implementation detail hidden in jenkins.model.Nodes, not sure how to expose it.

**Comment by Vlatombe** [comment]
Superceded by #2512

**Pull request closed without merging**